omod/src/main/webapp/pages/results.gsp

```diff
@@ -1,75 +1,163 @@
+                }
+            },
+            select: function () {
+                resultTable.columns(2).search(jq(this).val()).draw();
+            }
+        });
+
+        jq('#free-text-search').keyup(function(){
+            resultTable.search(jq(this).val()).draw();
+        });
+
+        jq('#reset-filters').click(function(){
+            // Reset the table and clear all the filters
+            resultTable.search('').columns().search('').draw();
+            jq('#rule-filter').prop('selectedIndex', 0);
+            jq('#patient-filter').val("");
+            jq('#free-text-search').val("");
+        });
+
     });
     <% } %>
 </script>
-<% if (rule == null) {
-%>
-<h2>${ui.message("dataintegrity.result.all.label")}</h2>
+
+<% if (binding.variables.containsKey('rules')) { %>
+    <h2>${ui.message("dataintegrity.result.all.label")}</h2>
 <% } else { %>
-<h2>Results for ${rule.ruleName}</h2>
+    <h2>Results for ${rule.ruleName}</h2>
 <% } %>
+
+<div id="filter-area">
+    <% if (binding.variables.containsKey('rules')) { %>
+    <label for="rule-filter">Rule:
+        <select id="rule-filter">
+            <option value="">Choose rule</option>
+            <% rules.each { %>
+            <option value="rule_${it.ruleId}">${ui.encodeHtmlContent(it.ruleName)}</option>
+            <% } %>
+        </select>
+    </label>
+    <% } %>
+    <label for="patient-filter">Choose Patient:
+        <input id="patient-filter">
+    </label>
+    <label for="free-text-search">Search:
+        <input id="free-text-search">
+    </label>
+    <button id="reset-filters">Reset</button>
+</div>
+
+
+
 <% if (results.size() == 0) { %>
     There are no results to show. <a href="results.page?action=fireRules"><i class="icon-play"></i>&nbsp;Click here</a> to run all the validation rules, please note that it will take some time and slow down the server.
 <% } else { %>
 <table id="list-results" cellspacing="0" cellpadding="2" width="100%" class="dataTable">
     <thead>
     <tr>
+        <th style="display: none"></th>
         <th>${ui.message("dataintegrity.rule.name")}</th>
-        <th>${ui.message("dataintegrity.result.id")}</th>
         <th>${ui.message("dataintegrity.result.name")}</th>
         <th>${ui.message("dataintegrity.result.notes")}</th>
         <th class="adminui-action-column">${ui.message("general.action")}</th>
     </tr>
     </thead>
     <tbody>
     <% results.each { %>
+    <script>
+        patients.push("${ui.encodeHtmlContent(it.name)}");
+    </script>
     <tr>
+        <td style="display: none">rule_${it.rule.ruleId}</td>
         <td>${ui.encodeHtmlContent(it.rule.ruleName ?: '')}</td>
-        <td>${it.patient.patientId ?: it.patientProgram.patientProgramId}</td>
-        <td>${ui.encodeHtmlContent(it.name ?: '')}</td>
+        <td>
+            <a href="/${contextPath}/coreapps/clinicianfacing/patient.page?patientId=${it.patient.uuid}">${ui.encodeHtmlContent(it.name ?: '')}</a>
```

Comment:
@shivtej1505 You also need have no link for the program but a link for the patient - remember the validation can be for either entity with more being added as time goes on